components/places/src/places.udl

```diff
@@ -0,0 +1,42 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+namespace places {};
+
+// Some kind of namespacing for uniffi would be ideal. Multiple udl/macro defns?
+// Everything below is from the crate::storage::history_metadata module...
+
+enum DocumentType {
+     // A page that isn't described by any other more specific types.
+    "Regular",
+    // A media page.
+    "Media",
+};
+
+// This is used as an "input" to the api.
+dictionary HistoryMetadataObservation {
```

Comment:
fwiw `apply_metadata_observation_impl` is basically already written against a `HistoryMetadataCompoundKey` type, except for the first few lines

Reply:
> I'd consider lifting (url, referrer, search_term) tuple into a `key` type, otherwise `url` as the only required item appears to act as an "anchor" here.

My take is that the key is an implementation detail but not strictly relevant to consumers. I'm also not convinced it is an implementation detail that will survive - it's seems like a reasonable implementation choice for the very first experiments, but it's not a fundamental property of anything. Future, more nuanced implementations might make different choices at different times - that it may end up turning into a "kind of a key, except when ..."

> I'd consider lifting (url, referrer, search_term) tuple into a `key` type, otherwise `url` as the only required item appears to act as an "anchor" here.

But that's entirely true. Url is the only required item. Url can not be blank, the other items can. When making an observation you should supply what you know. There's no natural concept of a "key" when visiting a page - it's just an artificial construct supporting the current implementation. The existing `VisitObservation` API doesn't specify the key, even though it's ending up in a DB and certainly will use a key. When querying history you also don't get to specify a "key" - you can query by various things, not just the database's primary key. I think the same thing applies here.

So if it's OK with you, I think I'd still rather leave `key` out at this time.

Reply:
Let's leave it out for now. I'm inclined to agree with your assessment of how this may play out in the future, but also inclined to think that the timelines will be quite stretched out (meaning, the current "key" shape will persist for quite a while). Let's see how it plays out!